### ISSUE TITLE
Implement `timeout()` within test block

### ISSUE THREAD

**Issue by Electroid** [oven-sh/bun#2030]
This [test](https://github.com/expressjs/express/blob/442fd467992992558806da8da07e945838712587/test/Route.js#L16) from `express` can't run because it need to define a custom timeout.

```ts
16 |   it('should not stack overflow with a large sync stack', function (done) {
17 |     this.timeout(5000) // long-running test
        ^
TypeError: undefined is not an object (evaluating 'this.timeout')
```

Also need to implement a default timeout for each block/file. We should match whatever Jest does.

**Comment by ThatOneBro** [comment]
So basically this is two parts:
* [ ] Add global timeout to test runner
* [ ] Add configurable local per-test timeout

**Comment by blackmann** [comment]
`this.timeout` is a mocha specific API, do we want to support that?

This commit implements timeout that matches that of Jest: https://github.com/oven-sh/bun/commit/7cf674a253862188c8c65057344a8ae5130413bb

Does it close this?

**Comment by Electroid** [comment]
We talked about this, will stick with Jest's version of timeouts for now. (which was implemented in https://github.com/oven-sh/bun/commit/7cf674a253862188c8c65057344a8ae5130413bb)

```js
import { test } from "bun:test";
import { sleep } from "bun";

test("should timeout", async () => {
  await sleep(100);
}, 10); // timeout is 10ms
```

**Comment by OrkhanAlikhanov** [comment]
Any way to change the default 5m to say 5 seconds globally like jest which has [5 seconds](https://jestjs.io/docs/configuration#testtimeout-number) by default?

**Comment by Jarred-Sumner** [comment]
> Any way to change the default 5m to say 5 seconds globally like jest which has [5 seconds](https://jestjs.io/docs/configuration#testtimeout-number) by default?

It's 5 seconds by default, not 5 minutes

https://github.com/oven-sh/bun/blob/25553e62c1cb8151c901f2e09ccb1a71f6a1e7bd/src/cli.zig#L946

**Comment by OrkhanAlikhanov** [comment]
Oh cool. I read 5 mins at [this line](https://github.com/oven-sh/bun/commit/7cf674a253862188c8c65057344a8ae5130413bb#diff-03a81e374ac6268f0e975230ed39affab3e807b987000fc5c8fd513359a82fd7R148)

**Comment by OrkhanAlikhanov** [comment]
It looks like there is `--timeout` option too. I didn't see that in docs for `bun test`.

**Comment by manast** [comment]
> We talked about this, will stick with Jest's version of timeouts for now. (which was implemented in [7cf674a](https://github.com/oven-sh/bun/commit/7cf674a253862188c8c65057344a8ae5130413bb))

It would be great to implement this feature now, as it seems like one of the few incompatibilities with Mocha, which is a hugely popular test framework, I am show-stopped by this issue trying to test BullMQ with Bun...

https://github.com/taskforcesh/bullmq/blob/master/tests/test_child-pool.ts#L88C7-L88C7

**Comment by endel** [comment]
In order to run `mocha` using Bun's runtime, I had to do the following:

```bash
bun run --bun node_modules/mocha/lib/cli/cli.js test/*Test.ts
```

Also had to manually update the `cli.js` file, due to this error: `"TypeError: yargs is not a function. (In 'yargs()', 'yargs' is an instance of Module)"`:

```diff
- const yargs = require('yargs/yargs');
+ const yargs = require('yargs');
```

Not sure if there's a better way of running `mocha` directly - I noticed a few failed test cases were missing in the output. Decided to document this in case someone else runs into the same in the future.
